### PR TITLE
Feature: Support `since` and `until` in supervisor journald wrapper

### LINKED ISSUE #2083: Support `since` and `until` in Balena Supervisor API
The currently implemented API endpoint contains a [subset](https://www.balena.io/docs/reference/supervisor/supervisor-api/#journald-logs) of the `journalctl` functionality. To retrieve logs within a closed time interval without manually filtering all retrieved logs, support for `since` and `until` is welcome.

### PR DESCRIPTION
**Pull request by rkeulemans**
# Description
The currently implemented API endpoint contains a [subset](https://www.balena.io/docs/reference/supervisor/supervisor-api/#journald-logs) of the journalctl functionality. To retrieve logs within a closed time interval without manually filtering all retrieved logs, support for `since` and `until` is welcome.

Fixes #2083

# Type of change
Change-type: minor

# Notes
- The `since` timestamp was already supported at the library side in `src/lib/journald.ts`, only the argument was not passed from the request body in `src/device-api/v2.ts`. At the library side the type of the `since` timestamp is defined to be a `number`. Therefore, it is assumed that this API was designed to take a UNIX timestamp as input. Since no type checking is happening at runtime, users could theoretically also pass any other argument valid to the javascript `Date()` constructor, if this is the desired API, the documentation should be adjusted accordingly. 

# How Has This Been Tested?
Tested with a device registered in Balena connected to the local network, using [these instructions](https://github.com/balena-os/balena-supervisor#developing-using-a-production-image-or-device).

### CODE REVIEW

File: docs/API.md
Comment (cywang117) [inline docs/API.md:1301]: This will need to be modified to the exact patch semver once this PR is ready to merge.

**Review by cywang117 — COMMENTED** [review]
Thanks for the PR @rkeulemans, and for adding tests for it too. Nice work following the contribution guidelines. I have no objections against adding these fields. However, I would advocate for `req.body.since` and `req.body.until` to have the same time format as what is expected by `journalctl`, and have the response error with a code 400 otherwise. Since the Supervisor API endpoint for journal is just a wrapper, it's better to avoid adding more complexity with time conversions.

**Comment by rkeulemans** [comment]
> Thanks for the PR @rkeulemans, and for adding tests for it too. Nice work following the contribution guidelines. I have no objections against adding these fields. However, I would advocate for `req.body.since` and `req.body.until` to have the same time format as what is expected by `journalctl`, and have the response error with a code 400 otherwise. Since the Supervisor API endpoint for journal is just a wrapper, it's better to avoid adding more complexity with time conversions.

@cywang117 Agreed! For any of the other currently supported flags, a HTTP status code `200` is returned, even if the provided value is not supported by `journalctl`. My proposal would be not to introduce any additional (duplicate) input parsing/validation and corresponding status codes, and leave the responsibility of providing the correct values to `journalctl` to the client, corresponding to the current implementation.

**Comment by cywang117** [comment]
Thanks for the chase @rkeulemans. I'll take a look at this tomorrow!

**Comment by cywang117** [comment]
@rkeulemans I think the `412` is a good idea in theory, and I'd totally be in support of it, but unfortunately we've already documented this API interface, and even if incorrect, Supervisor API consumers may already have code depending on the `200` status always being sent. In terms of interface changes, Supervisor devs had a discussion about this recently in response to [this ongoing PR](https://github.com/balena-os/balena-supervisor/pull/2048#discussion_r1065205210), and we agreed that the only interfaces that can change are messages attached to responses. 

This PR looks good to go though, once you change `14.5` in the docs to `14.5.0`. Let me know once you've done so and I can approve the merge. Thank you for your work!

**Comment by rkeulemans** [comment]
@cywang117 thanks for the quick follow-up! Just updated the version in the docs to `14.5.0`!

File: docs/API.md
Comment (cywang117) [inline docs/API.md:1295]:
```suggestion
> **Introduced in supervisor v14.5.0**
```

File: docs/API.md
Comment (cywang117) [inline docs/API.md:1300]:
```suggestion
> **Introduced in supervisor v14.5.0**
```

**Review by cywang117 — CHANGES_REQUESTED** [review]
LGTM after the changes!

**Review by cywang117 — APPROVED** [review]

**Comment by cywang117** [comment]
@rkeulemans , Apologies as I missed this last time, but would you mind squashing your commits into one to clean up the messages a bit? Thank you again!

**Review by cywang117 — CHANGES_REQUESTED** [review]
@rkeulemans , Apologies as I missed this last time, but would you mind squashing your commits into one to clean up the messages a bit? Thank you again!

**Comment by rkeulemans** [comment]
> @rkeulemans , Apologies as I missed this last time, but would you mind squashing your commits into one to clean up the messages a bit? Thank you again!

Did this just now!

**Comment by rkeulemans** [comment]
@cywang117 @pipex The CI is failing, do you have any suggestions on what is wrong?

**Review by cywang117 — APPROVED** [review]
Thank you @rkeulemans , approved!

EDIT: Regarding CI, I believe the `Change-type: minor` footer needs to be in the commit body, like so:

```
COMMIT MESSAGE HERE

Signed-off-by: your name <your@email.com>
Change-type: minor
Closes: https://github.com/balena-os/balena-supervisor/issues/2083
```

**Comment by rkeulemans** [comment]
@cywang117 Any clue why `Test npm (16.x)` is failing, the reported file `src/logging/monitor.ts` is not even touched in this PR?

File: src/lib/journald.ts
Comment (cywang117) [inline src/lib/journald.ts:14]:
```suggestion
	since?: number;
	until?: number;
```
@rkeulemans , I believe that would be because of the type change from number to string - you should be able to fix it by change the types of `since` and `until` back to number. The `args.join(' ')` call near the end of `spawnJournalCtl` will normalize the types. The specific error is because the `backfill` method in `src/logging/monitor.ts` passes a number type to as an opt for `since`.

**Comment by rkeulemans** [comment]
@cywang117 Thanks. Should be OK now!

**Comment by cywang117** [comment]
@balena-ci rebase

**Comment by cywang117** [comment]
@rkeulemans Our CI had a setting which disallowed some required checks from running, which I've disabled and merged into master. It looks like I can't rebase external contributions (which makes sense 😅 ) so could you rebase on top of master to see if the checks all run now? Thank you!

**Comment by rkeulemans** [comment]
@cywang117 Some tests are failing: `Error: Unsure how to proceed with event: pull_request_target`. Could you have a look?

**Comment by cywang117** [comment]
Hi @rkeulemans , I patched an outdated dependency in our CI/CD and all the tests run fine now. I've approved this so it should merge once you rebase and all the checks pass. Thanks for your patience!

**Review by cywang117 — APPROVED** [review]

**Comment by rkeulemans** [comment]
@cywang117 Still the policy-bot is pending, could you check again?

**Review by cywang117 — APPROVED** [review]

**Comment by cywang117** [comment]
@rkeulemans All good! Congrats on your first Supervisor contribution 🎉

**Pull request merged**
